lib/multi_patient_api/bulk_data_utils.rb

```diff
@@ -0,0 +1,383 @@
+			retry_after_val = retry_after.nil? || retry_after.value.nil? ? 0 : retry_after.value.to_i
+			wait_time = retry_after_val.positive? ? retry_after_val : wait_time *= 2
+
+			timeout -= Time.now - start + wait_time
+			sleep wait_time
+
+		end while response[:status] == 202 and timeout > 0
+
+	end 
+
+	# TODO: Delete this once core functionality is merged in
+	def stream(block, url = '', name: nil, **options)
+		store_request('outgoing', name) do
+			Faraday.get(url, nil, options[:headers]) { |req| req.options.on_data = block }
+		end
+	end 
+
+	def get_file(endpoint, use_token = true)
+		headers = { accept: 'application/fhir+ndjson' }
+		headers.merge!({ authorization: "Bearer #{bulk_access_token}" }) if use_token
+
+	 	get(endpoint, headers: headers)
+	end 
+
+	def stream_ndjson(endpoint, headers, process_chunk_line, process_response)
+
+		hanging_chunk = String.new 
+
+		process_body = proc { |chunk| 
+			
+			hanging_chunk << chunk 
+			chunk_by_lines = hanging_chunk.lines
+
+			hanging_chunk = chunk_by_lines.pop || String.new
+
+			chunk_by_lines.each do |elem| 
+				process_chunk_line.call(elem)
+			end 
+		}	
+
+		stream(process_body, endpoint, headers: headers)
+		process_chunk_line.call(hanging_chunk)
+
+		# TODO --> Would lijke for this block to get called once during the 
+		#					 block above so that we can check what the response is 
+		#					 and opt out if its bad or if the headers aren't what we
+		#					 need.
+		process_response.call(response) 
+
+	end
+
+	def predefined_device_type?(resource)  
+		return false if resource.nil?
+
+	end 
+
+	# TODO: Deal with device and lab edge cases
+	# Returns the canonical url denoting the profile
+	def determine_profile(profile_definitions, resource)
+
+		return profile_definitions[0][:profile] unless profile_definitions[0][:profile].nil?
+
+		assert false, "Profile for #{resource.class.name.demodulize} could not be determined." 
+
+		#return nil if resource.resourceType == 'Device' && !predefined_device_type?(resource)
+		#return nil if NON_US_CORE_KLASS.include(resource.resourceType)
+
+	end 
+
+	def resolve_element_from_path(element, path)
+		el_as_array = Array.wrap(element)
+		if path.empty?
+			return nil if element.nil?
+
+			return el_as_array.find { |el| yield(el) } if block_given?
+
+			return el_as_array.first
+		end
+
+		path_ary = path.split('.')
+		cur_path_part = path_ary.shift.to_sym
+		return nil if el_as_array.none? { |el| el.send(cur_path_part).present? || el.send(cur_path_part) == false }
+
+		el_as_array.each do |el|
+			el_found = if block_given?
+									 resolve_element_from_path(el.send(cur_path_part), path_ary.join('.')) { |value_found| yield(value_found) }
+								 else
+									 resolve_element_from_path(el.send(cur_path_part), path_ary.join('.'))
+								 end
+			return el_found if el_found.present? || el_found == false
+		end
+
+		nil
+	end
+
+	def find_slice_by_values(element, values)
```

Comment:
I am hoping that all of this must support checking can be handled by the `USCore` must support checking code.